libflux/src/core/semantic/convert.rs

```diff
@@ -138,6 +142,97 @@ fn convert_builtin_statement(stmt: ast::BuiltinStmt, fresher: &mut Fresher) -> R
     })
 }
 
+pub type MonoTypeMap = SemanticMap<String, MonoType>;
+#[allow(unused)]
+fn convert_monotype(
+    ty: ast::MonoType,
+    tvars: &mut HashMap<String, types::Tvar>,
+    f: &mut Fresher,
+) -> Result<MonoType> {
+    match ty {
+        ast::MonoType::Tvar(tv) => {
+            let tvar = tvars.entry(tv.name.name).or_insert_with(|| f.fresh());
+            Ok(MonoType::Var(*tvar))
+        }
+        ast::MonoType::Basic(basic) => match basic.name.name.as_str() {
+            "bool" => Ok(MonoType::Bool),
+            "int" => Ok(MonoType::Int),
+            "uint" => Ok(MonoType::Uint),
+            "float" => Ok(MonoType::Float),
+            "string" => Ok(MonoType::String),
+            "duration" => Ok(MonoType::Duration),
+            "time" => Ok(MonoType::Time),
+            "regexp" => Ok(MonoType::Regexp),
+            "bytes" => Ok(MonoType::Bytes),
+            _ => Err("Bad parameter type.".to_string()),
+        },
+        ast::MonoType::Array(arr) => Ok(MonoType::Arr(Box::new(types::Array(convert_monotype(
+            arr.element,
+            tvars,
+            f,
+        )?)))),
+        ast::MonoType::Function(func) => {
+            let mut req = MonoTypeMap::new();
+            let mut opt = MonoTypeMap::new();
+            let mut _pipe = None;
+            let mut dirty = false;
+            for param in func.parameters {
+                match param {
+                    ast::ParameterType::Required { name, monotype, .. } => {
+                        req.insert(name.name, convert_monotype(monotype, tvars, f)?);
+                    }
+                    ast::ParameterType::Optional { name, monotype, .. } => {
+                        opt.insert(name.name, convert_monotype(monotype, tvars, f)?);
+                    }
+                    ast::ParameterType::Pipe { name, monotype, .. } => {
+                        if !dirty {
+                            _pipe = Some(types::Property {
+                                k: match name {
+                                    Some(n) => n.name,
+                                    None => String::from("<-"),
+                                },
+                                v: convert_monotype(monotype, tvars, f)?,
+                            });
+                            dirty = true;
+                        } else {
+                            return Err("Bad parameter type.".to_string());
```

Comment:
Can you also use a better error message here? Like `"function types can have at most one pipe parameter"`